### DISCUSSION TITLE
Есть ли возможность получить название играющего инструмента стандартного банка?

### DISCUSSION THREAD

**Discussion by laika65** [melanchall/drywetmidi#173] [Q&A] [answered]
Сабж.

**Comment by melanchall** [comment, marked as answer]
Просто так в вакууме нельзя. Поэтому, пожалуйста, конкретизируйте, при каких обстоятельствах вы хотите получить инструмент?

Вы запускаете воспроизведение MIDI файла? Если так, то просто найдите события смены программы в нём. Если таких событий в файле нет, то возможности узнать инструмент тоже нет, всё зависит от синтезатора, сам стандарт MIDI не говорит, что считать дефолтной программой.

**Reply by laika65** [reply]
То есть мне надо, делать примерно следующие:
Заводить енум или массив типа такого
 ```csharp

        public string[] instrumentNames = {
            "AcousticGrandPiano",
            "BrightAcousticPiano",
            "ElectricGrandPiano",
            "HonkyTonk",
            // остальное
        };
```
и соотнести с ProgramChangeEvent?
А что за MidiEventType.InstrumentName? 
И ещё вопрос не по теме, как вытащить первые ProgramChangeEvent для каждого трека? В LINQ только вникаю.

**Reply by melanchall** [reply]
Не очень понятно всё-таки. Вы хотите получить какое-то константное значение программы из `ProgramChangeEvent`? Если вам нужна GM программа, вы можете сделать так:

```csharp
var program = (GeneralMidiProgram)(int)programChangeEvent.ProgramNumber;
```

Енум `GeneralMidiProgram` уже есть в библиотеке, если это то, что вам нужно.

_Instrument Name_ это такое мета-событие в стандарте SMF, туда в виде строки записывается имя инструмента. Это по сути произвольная строка, на неё ориентироваться не стоит, там вполне может быть что-то вроде `"My new super guitar"`.

По последнему вопросу:

```csharp
foreach (var trackChunk in midiFile.GetTrackChunks())
{
    var events = trackChunk.GetTimedEvents().Where(e => e.Event.EventType == MidiEventType.ProgramChange).ToArray();
}
```

`events` это будет массив [TimedEvent](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Interaction.TimedEvent.html).

**Comment by laika65** [comment]
Извиняюсь за долгий ответ, не было возможности написать. Всё получилось, но есть маленькая проблемка, которую пока не могу решить. Попробую сегодня ещё поломать голову и повникать в миди формат и если ничего не созреет в голове, то спрошу вас.

**Reply by melanchall** [reply]
Удалось решить проблему?

**Reply by laika65** [reply]
Отчасти, вот на видео в 3-ем файле проблемка:
https://youtu.be/XPaYMMY-5Xo
Почему то по трекам пробегается и заливает канвас разными нотами, но в списке всего один трек с нотами, надо вникнуть в миди события. Пока изучаю.
Сам файл, вдруг захотите взглянутъ.
[Lunarwind.zip](https://github.com/melanchall/drywetmidi/files/8549326/Lunarwind.zip)

**Reply by laika65** [reply]
Ещё вопрос не по теме. У меня 25 плейбэков, один основной и три группы по 8, 4 плейбэка загоняю в PlaybackCurrentTimeWatcher для отображения времени. Вот так это нормально реализовывать?
```csharp
            _timeWatcher.CurrentTimeChanged += (_, e) =>
            {
                var times = e.Times;
                foreach (var t in times)
                {
                    if (t.Playback == PlaybackMain)
                    {
                    }
                    else if (t.Playback == PlayerVM.Playback)
                    {
                    }
//остальные если ифы
                }
            };
```

**Reply by melanchall** [reply]
По последнему вопросу: да, вполне себе.

А касательно видео, тут нужны пояснения, ибо я не знаю ведь, что всё это означает :) Файл проверил, да, там только в одном чанке есть ноты. А из видео проблема не ясна. Рисуются ноты, всё нормально. Разный цвет в вашей программе = разные чанки? Если так, полагаю, проблема в отрисовке и с MIDI не связана.

**Reply by melanchall** [reply]
К слову, если колонка после колонки с названием инструмента показывает кол-во событий, то числа у вас какие-то странные там, DryWetMIDI отдаёт такие значения:

![image](https://user-images.githubusercontent.com/26361185/164970457-5cccff71-f30b-4cfb-bedd-ceac932fea97.png)

**Reply by laika65** [reply]
Информацию для датагрида я беру вот так:
```csharp
            foreach (var trackChunk in OpenedFile.GetTrackChunks())
            {
                foreach (var e in trackChunk.GetTimedEvents())
                {
                }
            }
```
От этого и разница в кол-ве, но это мне не особо важно. Больше интересует почему название инструмента такое. И я так понял это версия миди старая где инструменты меняются поcтоянно в одном треке событием ProgramChange, и видимо бесполезно получать названия инструмента в таких файлах.
Я всё время думал что треки в миди файлах это каналы, канвас заполняю так:
```csharp
            var notes = OpenedFile.GetNotes();
            foreach (Melanchall.DryWetMidi.Interaction.Note note in notes)
            {
                NoteItem rec = new NoteItem();
                switch (note.Channel)
                {
                    case 0:
                        rec.Background = Ch01Color;
                        break;
                    case 1:
                        rec.Background = Ch02Color;
                        break;
                        //остальные каналы
                }
                rec.Background.Freeze();
                Application.Current.Dispatcher.Invoke(
                  (Action)(() =>
                  {
                      CanvasNotes.Add(rec);
                  }), DispatcherPriority.Normal);
            }
```
И думал что трек чанки это каналы, но в таких файлах это не так, хотя в других всё нормально работает. Плеер меня не интересует совсем, его набросал, что бы понимать о структуре файлов и выводе нот на канвас. 
Спасибо большое за ответы и поддержку.

**Reply by melanchall** [reply]
> От этого и разница в кол-ве, но это мне не особо важно.

Хмм, а вот мне очень даже важно. Ибо если так, то это баг, проверю.

> Я всё время думал что треки в миди файлах это каналы

Вовсе нет, стандарт MIDI это никак не регламентирует. Так **обычно** делают, но это не непреложная истина, которой нужно следовать. В одном чанке спокойно могут быть канальные события хоть на всех 16 каналах.

И да, в таком случае, в чанке может вполне быть множество событий `ProgramChangeEvent`, ибо нужно для каждого канала выставить программу. Это во-первых. А во-вторых программа может меняться в течение файла даже для одного канала, тут тоже криминала нет.

**Reply by melanchall** [reply]
Нет, в библиотеке бага нет. Вы, видимо, как-то хитрее эти числа показываете.

![image](https://user-images.githubusercontent.com/26361185/164973358-ceafd58a-f58a-4f97-98fc-0ba03789fd63.png)

**Reply by melanchall** [reply]
Вот цитата из спецификации SMF (Standard MIDI Files):

> A track chunk contains a sequential stream of MIDI data which may contain information for up to 16 MIDI channels.

Т.е. стандарт просто говорит, что в общем случае в трек чанке содержатся MIDI-данные на 16 каналах.

Вообще, я рекомендую поизучать официальный документ, если интересно. Его можно [взять с сайта midi.org](https://www.midi.org/specifications-old/item/the-midi-1-0-specification), и вот тут ещё приложу: [complete_midi_96-1-3.pdf](https://github.com/melanchall/drywetmidi/files/8549533/complete_midi_96-1-3.pdf).

**Reply by laika65** [reply]
Да там такой огород, что стыдно показывать, потихоньку разбираюсь. 
Это у вас CLion? 
Я там и читаю, но всего много, что читаю, книжку по шарпу, док по вашей библиотеке, по миди, по WPF. Уже голова пухнет, хотя уже полегче. Сначала на решении чего-нибудь бывало тратил дни, сейчас быстро как то всё

**Reply by melanchall** [reply]
Не, это Visual Studio 2022, как и у вас, судя по видео :)
